### DISCUSSION TITLE
6.4 - Where to put our AppProviders?

### DISCUSSION THREAD

**Discussion by kevinfairclough** [remix-run/react-router#9424] [Q&A]
Hi

We have quite a few AppProviders where do we put them ? Around the RouterProvider or inside the RouterProvider ?

And how do we access data from contexts in loaders ? 

Thanks
Kevin

**Comment by supersnager** [comment]
I think your question is related to https://github.com/remix-run/react-router/discussions/9327

**Reply by kevinfairclough** [reply]
Thanks for the reply, Yep very similar

**Comment by ryanflorence** [comment]
If the provider doesn't need router/data information, you can put it on the outside:

```jsx
<SomeProvider>
  <RouterProvider router={router} />
</SomeProvider>
```

Or if the context needs routing/data information, put it in the root route element.

```jsx
const router createBrowserRouter([
  path: '/',
  element: <Root />
  children: [/* ... */]
]);

function Root() {
  const data = useLoaderData();
  return (
    <SomeProvider data={data}>
      <Outlet />
    </SomeProvider>
  )
}
```

**Reply by kevinfairclough** [reply]
Thanks for the reply, yeah we have metadata that needs to be loaded before the app can function.  The parallel loaders are great in the main however I need the root loader to wait then the rest can be in parallel.

**Reply by supersnager** [reply]
@kevinfairclough Maybe my solution will be helpful to you.
I'm using it to defer loading data until I'm sure the auth token exists.
You can return a Promise from the loader and resolve it when your metadata is ready.
But how do you know the metadata is ready if the loader can't use hooks?
It's simple - use Redux. Pass a store instance to the loader (you need a function that creates a loader), or import it and subscribe to the store:

```js
import store from "./myReduxStore";

const dataLoader = () => {
  return new Promise((resolve) => {
    const unsubscribe = store.subscribe( async () => {
    const currentState = store.getState();
     if (current.state.metadata) {
      const data = await loadData();
      // don't forget to unsubscribe
      unsubscribe();
      resolve(data);
     }
  })
  )
}
```
Of course, the metadata needs to be put into the store when it is ready.

Instead of Redux you can you any public-subscribe library or even native custom events.
